metrics/query.global/global.go

```diff
@@ -0,0 +1,194 @@
+	DEFAULT_PERCENTILE_OPTION     = "999"
+
+	BASE_QUERY     = "SELECT ROUND(bucket_quantile * 100, 1) AS p, ROUND(bucket_timer_high / 1000000, 3) AS us FROM performance_schema.events_statements_histogram_global"
+	TRUNCATE_QUERY = "TRUNCATE TABLE performance_schema.events_statements_histogram_global"
+)
+
+type Global struct {
+	db *sql.DB
+	// --
+	optional      map[string]bool      // keyed on level
+	percentiles   map[string][]float64 // keyed on level
+	truncateTable map[string]bool      // keyed on level
+	available     bool
+}
+
+var _ blip.Collector = &Global{}
+
+func NewGlobal(db *sql.DB) *Global {
+	return &Global{
+		db:            db,
+		optional:      map[string]bool{},
+		percentiles:   map[string][]float64{},
+		truncateTable: map[string]bool{},
+		available:     true,
+	}
+}
+
+// Domain returns the Blip metric domain name (DOMAIN const).
+func (c *Global) Domain() string {
+	return DOMAIN
+}
+
+// Help returns the output for blip --print-domains.
+func (c *Global) Help() blip.CollectorHelp {
+	return blip.CollectorHelp{
+		Domain:      DOMAIN,
+		Description: "Collect metrics for all queries",
+		Options: map[string]blip.CollectorHelpOption{
+			OPT_RESPONSE_TIME_OPTIONAL: {
+				Name:    OPT_RESPONSE_TIME_OPTIONAL,
+				Desc:    "If collecting response time metrics is optional. This will fail if it is required but not available",
+				Default: "yes",
+				Values: map[string]string{
+					"yes": "Optional",
+					"no":  "Required",
+				},
+			},
+			OPT_RESPONSE_TIME_PERCENTILES: {
+				Name:    OPT_RESPONSE_TIME_PERCENTILES,
+				Desc:    "Comma-separated list of response time percentiles formatted as 999, 0.999 or 99.9",
+				Default: DEFAULT_PERCENTILE_OPTION,
+				//Values:  map[string]string{},
+			},
+			OPT_TRUNCATE_TABLE: {
+				Name:    OPT_TRUNCATE_TABLE,
+				Desc:    "If the source table should be truncated to reset data after each retrieval",
+				Default: "yes",
+				Values: map[string]string{
+					"yes": "Truncate source table after each retrieval",
+					"no":  "Do not truncate source table after each retrieval",
+				},
+			},
+		},
+		Metrics: []blip.CollectorMetric{
+			{
+				Name: "response_time",
+				Type: blip.GAUGE,
+				Desc: "Query response time (microseconds)",
+			},
+		},
+	}
+}
+
+// Prepare prepares the collector for the given plan.
+func (c *Global) Prepare(ctx context.Context, plan blip.Plan) (func(), error) {
+	_, err := c.db.Query(BASE_QUERY)
+	if err != nil {
+		blip.Debug("response time metric is not available: %v", err.Error())
+		c.available = false
+	}
+
+LEVEL:
+	for _, level := range plan.Levels {
+		dom, ok := level.Collect[DOMAIN]
+		if !ok {
+			continue LEVEL // not collected at this level
+		}
+
+		err := c.prepareLevel(dom, level)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+// Collect collects metrics at the given level.
+func (c *Global) Collect(ctx context.Context, levelName string) ([]blip.MetricValue, error) {
+	if !c.available {
```

Comment:
I think you can just `return nil, nil` if not available. I need to double check but I'm pretty sure I set a bad example by checking `!c.optional[levelName]` in `Collect`, and now other devs have followed my bad example. The caller (`Monitor.Engine`) should only call `Collect` if and when it's correct to do so.

Reply:
The check here depends on the options specified by clients. If the metric is required but not available, we fail the collector with panic. Maybe we can do such check in `Prepare`?